packages/react-reconciler/src/ReactFiberScheduler.js

```diff
@@ -876,30 +876,41 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
     } while (true);
 
     // We're done performing work. Time to clean up.
-    stopWorkLoopTimer(interruptedBy);
-    interruptedBy = null;
-    isWorking = false;
+    let didCompleteRoot = null;
 
     // Yield back to main thread.
     if (didFatal) {
+      stopWorkLoopTimer(interruptedBy, didCompleteRoot);
+      interruptedBy = null;
+      isWorking = false;
       // There was a fatal error.
       return null;
     } else if (nextUnitOfWork === null) {
       // We reached the root.
       if (isRootReadyForCommit) {
+        didCompleteRoot = true;
+        stopWorkLoopTimer(interruptedBy, didCompleteRoot);
+        interruptedBy = null;
+        isWorking = false;
         // The root successfully completed. It's ready for commit.
         root.pendingCommitExpirationTime = expirationTime;
         const finishedWork = root.current.alternate;
         return finishedWork;
       } else {
         // The root did not complete.
+        stopWorkLoopTimer(interruptedBy, didCompleteRoot);
+        interruptedBy = null;
+        isWorking = false;
         invariant(
           false,
           'Expired work should have completed. This error is likely caused ' +
             'by a bug in React. Please file an issue.',
         );
       }
     } else {
+      stopWorkLoopTimer(interruptedBy, didCompleteRoot);
+      interruptedBy = null;
+      isWorking = false;
```

Comment:
Sorry, another nit: can we move `isWorking` back out? :) It's not related to the work timer and it's the same in every branch.